File: zalo_service/src/crypto/transport-cipher.ts
import {
  createCipheriv,
  createDecipheriv,
  createHash,
  randomBytes,
} from 'node:crypto';

import { loadConfig } from '../config.js';

/**
 * Mirror of Rails `Zalo::TransportCipher` — the two must stay byte-compatible.
 *
 * Zalo session cookies used to cross the Rails <-> sidecar boundary in
 * plaintext over loopback HTTP, so anything observing that traffic saw full
 * session credentials (red team finding C3). Both processes already share
 * ZALO_SERVICE_INTERNAL_TOKEN, so the key is derived from it rather than
 * introducing another secret to distribute.
 *
 * Wire format: base64( iv[12] || ciphertext || gcmTag[16] )
 */

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;
const TAG_LENGTH = 16;

function derivedKey(): Buffer {
  const token = loadConfig().ZALO_SERVICE_INTERNAL_TOKEN;
  return createHash('sha256').update(token, 'utf8').digest();
}

export function encryptForRails(plaintext: string): string {
  const iv = randomBytes(IV_LENGTH);
  const cipher = createCipheriv(ALGORITHM, derivedKey(), iv);
  const ciphertext = Buffer.concat([
    cipher.update(plaintext, 'utf8'),
    cipher.final(),
  ]);
  return Buffer.concat([iv, ciphertext, cipher.getAuthTag()]).toString(
    'base64',
  );
}

export function decryptFromRails(payload: string): string {
  const raw = Buffer.from(payload, 'base64');
  if (raw.length <= IV_LENGTH + TAG_LENGTH) {
    throw new Error('transport-cipher: payload too short to be valid');
  }

  const decipher = createDecipheriv(
    ALGORITHM,
    derivedKey(),
    raw.subarray(0, IV_LENGTH),
  );
  decipher.setAuthTag(raw.subarray(raw.length - TAG_LENGTH));
  return Buffer.concat([
    decipher.update(raw.subarray(IV_LENGTH, raw.length - TAG_LENGTH)),
    decipher.final(),
  ]).toString('utf8');
}

File: app/services/zalo/transport_cipher.rb
# Encrypts the credential fields that cross the Rails <-> Node sidecar boundary.
#
# Zalo session cookies are encrypted at rest by Active Record, but the internal
# API used to hand them to the sidecar in plaintext over loopback HTTP, so
# anything able to observe that traffic — tcpdump on lo, an intercepting proxy,
# a request log — saw full session credentials (red team finding C3).
#
# AES-256-GCM under a key derived from the shared service token both processes
# already hold. This closes passive capture of the wire. It is not a substitute
# for mTLS: an attacker who can already read either process's environment holds
# the token too.
class Zalo::TransportCipher
  class DecryptionError < StandardError; end

  CIPHER = 'aes-256-gcm'.freeze
  IV_LENGTH = 12
  TAG_LENGTH = 16

  class << self
    def encrypt(plaintext)
      return if plaintext.nil?

      cipher = OpenSSL::Cipher.new(CIPHER).encrypt
      cipher.key = key
      iv = cipher.random_iv
      ciphertext = cipher.update(plaintext) + cipher.final
      Base64.strict_encode64(iv + ciphertext + cipher.auth_tag)
    end

    def decrypt(payload)
      return if payload.blank?

      raw = Base64.strict_decode64(payload)
      cipher = OpenSSL::Cipher.new(CIPHER).decrypt
      cipher.key = key
      cipher.iv = raw[0, IV_LENGTH]
      cipher.auth_tag = raw[-TAG_LENGTH..]
      cipher.update(raw[IV_LENGTH...-TAG_LENGTH]) + cipher.final
    rescue ArgumentError, OpenSSL::Cipher::CipherError => e
      # A bad tag means the payload was truncated, corrupted or forged; a bad
      # Base64 means it was never ciphertext. Neither is recoverable.
      raise DecryptionError, e.message
    end

    private

    def key
      token = ENV.fetch('ZALO_SERVICE_INTERNAL_TOKEN', '')
      raise 'ZALO_SERVICE_INTERNAL_TOKEN must be set to exchange Zalo credentials with the sidecar' if token.blank?

      OpenSSL::Digest::SHA256.digest(token)
    end
  end
end
